File: ruby/find_shortest_string_recursive.rb
def find_shortest_string_recursive(arr)
  return arr.first if arr.length == 1

  result = find_shortest_string_recursive(arr.drop(1))

  if arr[0].length <= result.length
    return arr[0]
  else
    return result
  end
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 'a'"
  puts "=>", find_shortest_string_recursive(['aaa', 'a', 'bb', 'ccc'])

  puts

  puts "Expecting: 'hi'"
  puts "=>", find_shortest_string_recursive(['cat', 'hi', 'dog', 'an'])

  puts

  puts "Expecting: 'lily'"
  puts "=>", find_shortest_string_recursive(['flower', 'juniper', 'lily', 'dandelion'])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: javascript/find_shortest_string_recursive.js
function findShortestStringRecursive(arr) {
  if (arr.length === 1){
    return arr[0]
  }
  if (arr[0].length <= arr[1].length) {
    return findShortestStringRecursive([arr[0], ...arr.slice(2)])
  } else {
    return findShortestStringRecursive(arr.slice(1))
  }
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 't'");
  console.log("=>", findShortestStringRecursive(['t', 'act', 'bb', 'ccc']));

  console.log("Expecting: 'a'");
  console.log("=>", findShortestStringRecursive(['aaa', 'a', 'bb', 'ccc']));

  console.log("Expecting: 'hi'");
  console.log("=>", findShortestStringRecursive(['cat', 'hi', 'dog', 'an']));

  console.log("Expecting: 'lily'");
  console.log("=>", findShortestStringRecursive(['flower', 'juniper', 'lily', 'dandelion']));
}

module.exports = findShortestStringRecursive;

// Please add your pseudocode to this file
// And a written explanation of your solution
